kv/etcd/etcd.go

```diff
@@ -0,0 +1,292 @@
+			context.Background(),
+			c.opts.KeyFn()(key),
+			clientv3.WithProgressNotify(),
+			// Receive initial notification once the watch channel is created
+			clientv3.WithCreatedNotify(),
+		)
+		c.m.etcdWatchCreate.Inc(1)
+
+		watchable = kv.NewValueWatchable()
+		c.watchables[key] = watchable
+
+		go func() {
+			ticker := time.Tick(c.opts.WatchChanCheckInterval())
+
+			for {
+				select {
+				case r, ok := <-watchChan:
+					if ok {
+						c.processNotification(r, watchable, key)
+					} else {
+						c.logger.Warnf("etcd watch channel closed on key %s, recreating a watch channel", key)
+
+						// avoid recreating watch channel too frequently
+						time.Sleep(c.opts.WatchChanResetInterval())
+
+						watchChan = c.watcher.Watch(
+							context.Background(),
+							c.opts.KeyFn()(key),
+							clientv3.WithProgressNotify(),
+							clientv3.WithCreatedNotify(),
+						)
+						c.m.etcdWatchReset.Inc(1)
+					}
+				case <-ticker:
+					c.RLock()
+					numWatches := watchable.NumWatches()
+					c.RUnlock()
+
+					if numWatches != 0 {
+						// there are still watches on this watchable, do nothing
+						continue
+					}
+
+					if cleanedUp := c.tryCleanUp(key); cleanedUp {
+						return
+					}
+				}
+			}
+		}()
+	}
+	c.Unlock()
+	_, w, err := watchable.Watch()
+	return w, err
+}
+
+func (c *client) tryCleanUp(key string) bool {
+	c.Lock()
+	defer c.Unlock()
+	watchable, ok := c.watchables[key]
+	if !ok {
+		// not expect this to happen
+		c.logger.Warnf("unexpected: key %s is already cleaned up", key)
+		return true
+	}
+
+	if watchable.NumWatches() != 0 {
+		// a new watch has subscribed to the watchable, do not clean up
+		return false
+	}
+
+	watchable.Close()
+	delete(c.watchables, key)
+	return true
+}
+
+func (c *client) processNotification(r clientv3.WatchResponse, w kv.ValueWatchable, key string) {
+	err := r.Err()
+	if err != nil {
+		c.logger.Errorf("received error on watch channel: %v", err)
+		c.m.etcdWatchError.Inc(1)
+	}
+
+	// we need retry here because if Get() failed on an watch update,
+	// it has to wait 10 mins to be notified to try again
+	if err = c.retrier.Attempt(func() error {
+		return c.update(w, key)
+	}); err != nil {
+		c.logger.Errorf("received notification for key %s, but failed to get value: %v", key, err)
+	}
+}
+
+func (c *client) update(w kv.ValueWatchable, key string) error {
+	v, err := c.Get(key)
+	if err != nil {
+		return err
+	}
+
+	curValue := w.Get()
+	if curValue != nil && curValue.Version() >= v.Version() {
+		return nil
+	}
+
+	return w.Update(v)
+}
+
+func (c *client) Set(key string, v proto.Message) (int, error) {
+	ctx, cancel := c.context()
+	defer cancel()
+
+	value, err := proto.Marshal(v)
+	if err != nil {
+		return 0, err
+	}
+
+	r, err := c.kv.Put(ctx, c.opts.KeyFn()(key), string(value), clientv3.WithPrevKV())
+	if err != nil {
+		c.m.etcdPutError.Inc(1)
+		return 0, err
+	}
+
+	// if there is no prev kv, means this is the first version of the key
+	if r.PrevKv == nil {
```

Comment:
You can't set the same version with the Set API, if 2 request came in, the fast one will inc the version v0 to v0+1 first then the second one will inc it from v0+1 to v0+2, they will be able to see different previous versions(v0 and v0+1) and they will surely see them because they are both only 1 revision old. 

The CAS API is for setting the value based on a certain version, where the atomicity will be guaranteed by using transaction